Improve displayed version name when building from PR

abravalheri opened the pull request:
Hello RTD maintainers and thank you very much for the tool, it is really amazing.

This PR is actually an improvement suggestion:

Currently, the displayed version in the case of builds originated from pull requests is simply a plain integer corresponding to the PR number in GitHub/GitLab. This can be a bit confusing, specially for new projects, because the PR numbers will be in the same range as possibly existing versions.

Therefore the idea is to make it more explicit when building from PRs by adding a `PR-` prefix to the PR number (or `MR-` in the case of GitLab).

A few notes:
- You can see that there are some changes that were automatically introduced by `pre-commit` when I tried to run `git commit`. These format changes are not part of the original idea and not necessary to it. To facilitate distinguishing between the actual suggestion and the automatic changes I tried to split them into 2 separated commits (the second commit in this PR being the one corresponding to the changes created by `pre-commit`). There is one exception though regarding statements starting with `if all([`, which I had to manually tweak after the automatic code formatter, because the result was too weird and increased the number of errors/warnings returned by the linter.
- When I run `tox -e lint` there are a few error messages corresponding to pre-existent code
- When I run `tox`, there are a few failing tests that seem to be uncorrelated to the introduced changes. In fact, by checking other open PRs I can see that the same tests are failing elsewhere (for example [here](https://app.circleci.com/pipelines/github/readthedocs/readthedocs.org/1742/workflows/547f7979-916d-4657-acab-bd51d3133f09/jobs/3438)). They seem to be related to a change in API urls (or at least to the way they are being tested):
    ```diff
    @@  @@
    'urls': {
        'dashboard': {
    -       'edit': 'https://readthedocs.org/dashboard/project/version/v1.0/edit/',
    +       'edit': 'https://readthedocs.org/dashboard/project/version/v1.0/',
    @@  @@
    ```
- I originally thought about making the prefixes into something that cannot correspond to existing branch names and therefore very distinctive and not subject to name collisions, for example `~PR-#2` or `~MR-#2` (with the prefix being `"~PR-#"`/`"~MR-#"`; `~` is an [invalid branch name character](https://wincent.com/wiki/Legal_Git_branch_names)). However, since I don't know how the remaining of the RTD system would deal with these unconventional characters (maybe it is validated somewhere), I decided do simply leave it as `"PR-"`/`"MR-"`. Please let me know if the original idea is safe and you would prefer it that way.

stsewd commented on docs/webhooks.rst:
```diff
@@ -59,6 +59,8 @@ GitHub
 * Leave the **Secrets** field blank
 * Select **Let me select individual events**,
   and mark **Branch or tag creation**, **Branch or tag deletion** and **Pushes** events
+* If you are using Read the Docs to :doc:`preview documentation changes from pull
```
This can be a part of the above sentence (events, if you are using...)

stsewd reviewed (commented):
Hi, thanks for the PR, we depend on the verbose_name being the identifier in a few places, changing this means we need to change a couple of other places that depend on this.

> This can be a bit confusing, specially for new projects, because the PR numbers will be in the same range as possibly existing versions.

Do you have examples of projects that use just a number for versions? I think most projects use semver `1.x` instead of just `1`.


Where do you find showing the number confusing? I think we always put the number close to the text "Pull request #xxx".

abravalheri commented:
Thank you very much for the review @stsewd .

Could you point me out which other places the verbose_name is being used to identify the PR so I can also have a look and maybe improve? Of course if it is very complicated, please feel free to close this PR, in the end it is a matter of minor experience improvement. Unfortunately the whole setup for developing, specially when involving PRs was quite cumbersome for me, so I ended up just relying on the test suite.

The place I found it confusing was the version switch menu. The version shows up as a number (let's say `4`), togheter with the other versions. So let's say you end up with something like `v0, v0.1, v1, v1.1, v1.2, 4` in that menu.

After some thought we can figure it out due to the sequence continuity, but I think making it more explicit would be a nice improvement.

stsewd commented:
> The place I found it confusing was the version switch menu. The version shows up as a number (let's say 4), togheter with the other versions. So let's say you end up with something like v0, v0.1, v1, v1.1, v1.2, 4 in that menu.

I see, then I think we should implement that logic in our footer to make it more explicit that the version is the PR/MR identifier.

Maybe something like `1 (PR)` or `#1`

abravalheri commented:
That makes a lot of sense to me.

Is this string used in the footer the same as the one being set to `READTHEDOCS_VERSION`?
The reason why I tapped into the `verbose_name` is because I thought that was the one being used to set `READTHEDOCS_VERSION` and the footer.

Personally, I used the `READTHEDOCS_VERSION` to set `html_title = f"{project} {READTHEDOCS_VERSION}"` so would also be very happy if the env var is also more explicit.

Maybe if you point me in the right direction I could try to contribute with that? This was me first interaction with the code base, so not very sure on how to tackle this.

stsewd commented:
> Personally, I used the `READTHEDOCS_VERSION` to set `html_title = f"{project} {READTHEDOCS_VERSION}"` so would also be very happy if the env var is also more explicit.
> 
> Maybe if you point me in the right direction I could try to contribute with that? This was me first interaction with the code base, so not very sure on how to tackle this.

That env var is the slug of the version, not the name to be displayed, so we can't change that to include other text.

https://github.com/readthedocs/readthedocs.org/blob/a91bd99ed8f30e60b52d88bec9091e7fd4a181f7/readthedocs/projects/tasks.py#L326-L334

I think we can add a new env var like `READTHEDOCS_VERSION_TYPE` (here the type can be `branch`, `tag`, or `external` (that is for PRs), that way you could add a suffix/prefix when needed.


The footer can be changed here https://github.com/readthedocs/readthedocs.org/blob/a91bd99ed8f30e60b52d88bec9091e7fd4a181f7/readthedocs/api/v2/templates/restapi/footer.html#L41-L46

There you would need to check if the version is external https://github.com/readthedocs/readthedocs.org/blob/a91bd99ed8f30e60b52d88bec9091e7fd4a181f7/readthedocs/builds/models.py#L900-L906

and add the prefix/suffix.

If you want to check your changes locally, you can check our docs at https://docs.readthedocs.io/en/stable/development/install.html

abravalheri commented:
Thank you very much for the tips @stsewd.

I have rewritten the PR to reflect the discussion in the comments.

stsewd commented on readthedocs/api/v2/templates/restapi/footer.html:
```diff
@@ -6,7 +6,7 @@
     <span class="rst-current-version" data-toggle="rst-current-version">
       <span class="fa fa-book">&nbsp;</span>
-      v: {{ current_version }}
+      v: {{ version.explicit_name }}
```
Looks like we are passing `version` and `current_version` so it doesn't get shallowed in the loop, I think we can make this `version` instead of only having the `verbose_name`.

https://github.com/readthedocs/readthedocs.org-ghsa-3v5m-qmm9-3c6c/blob/56100454427561fd225b0315ac9b259eac0d9da6/readthedocs/api/v2/views/footer_views.py#L171-L171

abravalheri replied:
I think it should be more or less how you have described now.

stsewd commented on readthedocs/builds/models.py:
```diff
@@ -200,6 +199,43 @@ def __str__(self):
         return self.type == EXTERNAL
 
+    @property
+    def explicit_external_name(self):
```
I think we can merge this with the function below.

stsewd commented on readthedocs/api/v2/templates/restapi/footer.html:
```diff
@@ -42,7 +42,7 @@
         {% for version in versions %}
         <dd {% if version.verbose_name == current_version %} class="rtd-current-item" {% endif %}>
-          <a href="{{ version.get_subdomain_url }}{{ path|default_if_none:"" }}">{{ version.slug }}</a>
+          <a href="{{ version.get_subdomain_url }}{{ path|default_if_none:"" }}">{{ version.explicit_name }}</a>
```
ha, we were displaying the slug here, and the verbose name in the "current version", +1 on using explicit_name for both.

stsewd commented on readthedocs/builds/utils.py:
```diff
@@ -79,6 +86,33 @@ def get_vcs_url(*, project, version_type, version_name):
+    The abbreviation is formed for the first letter in each word in the string,
+    capitalized.
+    """
+    return ''.join(word[0].upper() for word in string.split())
```
I think this one can also be moved to where the function is used. But I'm also not sure what's best having `PR` or `Pull Request`

stsewd commented on docs/builds.rst:
```diff
@@ -80,11 +80,23 @@ The *Sphinx* and *Mkdocs* builders set the following RTD-specific environment va
 
+.. note::
+
+   - The term "slug" is used here to indicate a string without some types characters
+     considered special in certain circunstances. The version slug can contain some
+     characters not usually present in other slugs.
```
Good to describe the slug :+1:, I don't think we explain that in any other place p:

```suggestion
   - The term slug is used refer to a unique string across projects/versions containing ASCII characters only.
     This value is used in the URLs of your documentation.
```

stsewd commented on docs/builds.rst:
```diff
@@ -80,11 +80,23 @@ The *Sphinx* and *Mkdocs* builders set the following RTD-specific environment va
+     considered special in certain circunstances. The version slug can contain some
+     characters not usually present in other slugs.
+
+   - The ``external`` value for ``READTHEDOCS_VERSION_TYPE`` will usually be
+     associated with builds triggered by pull requests. See :doc:`/pull-requests`.
```
```suggestion
   - If ``READTHEDOCS_VERSION_TYPE`` is ``external``,
     it means that the version was built from pull request. See :doc:`/pull-requests`.
```

stsewd commented on readthedocs/projects/tasks.py:
```diff
@@ -317,6 +317,8 @@ def get_rtd_env_vars(self):
             'READTHEDOCS': 'True',
             'READTHEDOCS_VERSION': self.version.slug,
+            'READTHEDOCS_VERSION_TYPE': self.version.type,
+            'READTHEDOCS_VERSION_NAME': self.version.explicit_name,
```
```suggestion
            'READTHEDOCS_VERSION_NAME': self.version.verbose_name,
```

My thinking here was that users could get the original name and add any other information based on the version_type.

stsewd reviewed (commented):
Thanks for working on this!

abravalheri commented:
Thank you very much for the review and suggestions @stsewd, I submitted a new commit rebasing the new state of the master an adopting the changes you have described.

Regarding `#4 (PR)` x `#4 (Pull Request)`, I agree that the second is even more explicit and therefore it would be better.

I suppose the layouts for the RTD theme and the floating menu for the other themes handle well long version names, right? For example this PR would have a name `#8237 (Pull Request)`.
If that is fine, I can also change and remove the abbreviation. Maybe you want to discuss it with your colleagues? Please let me know and I will be happy to change that.

stsewd reviewed (approved):
Thanks! Looks good for me! /cc @readthedocs/core

ericholscher commented on readthedocs/builds/models.py:
```diff
@@ -200,6 +198,28 @@ def __str__(self):
+        For example, if a version originates from GitHub pull request #4, then
+        ``version.explicit_name == "#4 (PR)"``.
+
+        On the other hand, Versions associated with regular ReadTheDocs builds
```
```suggestion
        On the other hand, Versions associated with regular RTD builds
```

ericholscher commented on readthedocs/api/v2/templates/restapi/footer.html:
```diff
@@ -41,8 +41,8 @@
-        <dd {% if version.verbose_name == current_version %} class="rtd-current-item" {% endif %}>
-          <a href="{{ version.get_subdomain_url }}{{ path|default_if_none:"" }}">{{ version.slug }}</a>
+        <dd {% if version == current_version %} class="rtd-current-item" {% endif %}>
+          <a href="{{ version.get_subdomain_url }}{{ path|default_if_none:"" }}">{{ version.explicit_name }}</a>
```
I don't think this is needed -- we will never show a PR build in this list. The slug is what will be shown in the URL, so I think we want the slug?

stsewd replied:
We were showhing the name for the current version, and the slug for the other versions, I think it makes sense showing the verbose name as this is how versions are originally named.

ericholscher commented on readthedocs/builds/models.py:
```diff
@@ -200,6 +198,28 @@ def __str__(self):
+        """
+        external_origin = external_version_name(self)
+
+        if not external_origin:
```
We should check for `if not self.type == EXTERNAL` here or similar -- it's cleaner than depending on the side effects of another function.

ericholscher reviewed (commented):
This seems like a good change, but I'm a bit worried about the extension of the build context -- we are trying to move more away from using that data without it being a nicer UX. Why was that necessary for this PR?

abravalheri commented:
Thank you very much for the review @ericholscher. I rebased the PR to master and added some changes to address some parts of your comments.

Regarding your comment:

> This seems like a good change, but I'm a bit worried about the extension of the build context -- we are trying to move more away from using that data without it being a nicer UX. Why was that necessary for this PR?

I am not exactly sure if I understand it. What would it be the build context? Also not sure what it would mean to have "data without it being a nicer UX"... Sorry for that (it seems that you guys are used to some terms when talking about the project, but for me as an external contributor it is a bit difficult to understand).

ericholscher commented on docs/builds.rst:
```diff
@@ -80,11 +80,22 @@ The *Sphinx* and *Mkdocs* builders set the following RTD-specific environment va
 
    ``READTHEDOCS``, Whether the build is running inside RTD, ``True``
-   ``READTHEDOCS_VERSION``, The RTD name of the version which is being built, ``latest``
+   ``READTHEDOCS_VERSION``, The RTD slug of the version which is being built, ``latest``
```
> I am not exactly sure if I understand it. What would it be the build context? Also not sure what it would mean to have "data without it being a nicer UX"... Sorry for that (it seems that you guys are used to some terms when talking about the project, but for me as an external contributor it is a bit difficult to understand).

This is the build context I'm talking about. We have a [proposed plan](https://docs.readthedocs.io/en/latest/development/design/theme-context.html) to change this, but we're now not 100% sure we want additional info here. What's the goal of providing this data?

stsewd replied:
These are environment variable, not the html sphinx context. This is so users can read these vars and do things like `html_title = f"{project} {READTHEDOCS_VERSION}"` https://github.com/readthedocs/readthedocs.org/pull/8237#issuecomment-856092333 (having some extra logic if the version is external). Having this as env var doesn't break the current html context, and it works for all tools. We can omit this change from this PR if you consider, and leave only the change from the version name on the footer.

Another addition for env vars would be https://github.com/readthedocs/readthedocs.org/issues/8260

abravalheri replied:
Thank you very much @stsewd for explaining the need for the env vars.

@ericholscher indeed my main reason for exposing these variables is what @stsewd pointed out. I hope to use the information provided by RTD to compose suitable titles/automatic texts in the sphinx templates. Expanding a bit on the example, these env vars would allow something like the following:
```python
if READTHEDOCS_VERSION_TYPE:
	html_title = f"{project} PR #{READTHEDOCS_VERSION_NAME}"
else:
	html_title = f"{project} {READTHEDOCS_VERSION_NAME}"
```
(the difference between the slug and the verbose name being pointed out in https://github.com/readthedocs/readthedocs.org/pull/8237#discussion_r660676316)

They should not interfere at all with the sphinx build process, but rather help the documentation authors to properly display certain properties in the documentation.

humitos replied:
We are fine adding the environment variable here as this does not affect the Sphinx's build context and it's done in a generic and useful way.

pmeier commented:
Any update on this?

humitos commented:
@ericholscher can you take another look at this PR? It looks like a good addition but you have made some comments regarding the environment variable and build context. After a quick read of the comments, I think I'm fine adding the environment variable but not into the HTML context.

humitos commented on docs/webhooks.rst:
```diff
@@ -58,7 +58,9 @@ GitHub
 * Leave the **Secrets** field blank
 * Select **Let me select individual events**,
-  and mark **Branch or tag creation**, **Branch or tag deletion** and **Pushes** events
+  and mark **Branch or tag creation**, **Branch or tag deletion** and **Pushes** events.
+  If you want to use Read the Docs to :doc:`preview documentation changes from pull
+  requests </pull-requests>`, also mark **Pull requests** events.
```
```suggestion
  and mark **Branch or tag creation**, **Branch or tag deletion**, **Pushes** and **Pull requests** events.
```

I think it's better to just mark those without too many explanations. Otherwise, people will need to modify the webhook if they want that feature later.

ericholscher commented:
@humitos 

> After a quick read of the comments, I think I'm fine adding the environment variable but not into the HTML context.

Agreed. I think ENV vars are fine, but in some ways we're just changing around where the data is coming from. I don't want to keep adding things there without a bit of thought either on namespacing at least. I'm 👍 on exposing this data though if there isn't a good way to get it currently.

humitos commented on docs/builds.rst:
```diff
@@ -80,11 +80,22 @@ The *Sphinx* and *Mkdocs* builders set the following RTD-specific environment va
+   - The term slug is used to refer to a unique string across projects/versions containing ASCII characters only.
+     This value is used in the URLs of your documentation.
+
+   - If ``READTHEDOCS_VERSION_TYPE`` is ``external``,
+     it means that the version was built from pull request. See :doc:`/pull-requests`.
```
I'd say to move this text to the description of the variable itself instead of in a note. After that, we remove the bullets and keep just the initial paragraph of the note which is great.

humitos reviewed (commented):
I re-review this PR and I think it's good to be merged. I left just small improvements that should be easy to adjust.

@abravalheri do you have the time to update this PR so we can merge it?

abravalheri commented:
Hi @humitos, thank you very much for the latest reviews.
I tried to rebase, fix the conflicts and apply the latest changes you suggested. I hope it is good.

I can see that the `ci/circleci: checks` are failing and in their logs they mention `pre-commit`.
When I try to run the same command locally, pre-commit changes a bunch of stuff that was not touched by this PR... (some parts were touched, but the coding style adopted was in keep with the current code-style found in the files).

I can add another commit adding the automatic changes suggested by `pre-commit`, but I am afraid it will introduce a lot of unrelated changes that will cloud the diff in the PR...

humitos commented:
> I can add another commit adding the automatic changes suggested by `pre-commit`, but I am afraid it will introduce a lot of unrelated changes that will cloud the diff in the PR...

Yeah, this is unfortunate because we are in the process of "blackify" the whole codebase starting only with the files touched by each PR. I just checked running `tox -e pre-commit` by myself in your branch and I think it will be fine adding a commit with those changes.

The PR is already reviewed and I'd say it's ready to be merged. In any case, we can skip that particular commit if we want to re-re-re-view it again 😄 --but I don't think it will be needed.

I'm sorry for all the time and work this took to you and thanks for it.

abravalheri commented:
> think it will be fine adding a commit with those changes.

Perfect, I just submitted 2eab243 with the changes automatically generated by pre-commit.

I got the configuration file from https://github.com/readthedocs/common/blob/main/pre-commit-config.yaml

Hopefully that will be enough to fix the problems in the CI and not cause any unrelated error 😅 🤞.

ericholscher commented:
This should now be fixed, please let me know if it isn't.
